### PR TITLE
Content Layer - Step1 - Project Settings #7892

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerNotFoundException.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKey.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayers.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKeys.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  4
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerAlreadyExistsException.java  1
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/ContentLayerJson.java  9
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerName.java  4
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectJson.java  2
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/ProjectServiceImpl.java  2
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/DeleteLayerParamsJson.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  4
         

Complexity decreasing per file
==============================
+ modules/core/core-api/src/main/java/com/enonic/xp/project/ModifyProjectParams.java  -1
         

Clones added
============
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=5108750&bid=16799807)

File: modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+{
+    private final ContentLayerKey key;
+
+    private final List<ContentLayerKey> parentKeys;
+
+    private final String displayName;
+
+    private final String description;
+
+    private final Locale language;
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java]: please don't call locale `language`. It is confusing. Call it locale.

File: modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerConstants.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.enonic.xp.project.layer;
+
+public final class ContentLayerConstants
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerConstants.java]: should have private constructor

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKeys.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayers.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKey.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerName.java  4
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/DeleteLayerParamsJson.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerConstants.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerAlreadyExistsException.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  4
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectJson.java  2
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerNotFoundException.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/ProjectServiceImpl.java  2
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/ContentLayerJson.java  9
         

Complexity decreasing per file
==============================
+ modules/core/core-api/src/main/java/com/enonic/xp/project/ModifyProjectParams.java  -1
         

Clones added
============
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=5108750&bid=16799807)

File: modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    private final String description;
+
+    private final Locale locale;
+
+    private final Attachment icon;
+
+    private ContentLayer( final Builder builder )
+    {
+        Preconditions.checkNotNull( builder.key, "Name cannot be null" );
```
Comment (anatol-sialitski) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java]: Why a message for name instead of key?

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectJson.java  2
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/ProjectServiceImpl.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKeys.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerKey.java  4
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/ContentLayerJson.java  9
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayers.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerName.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  4
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/layer/json/DeleteLayerParamsJson.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayerConstants.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerNotFoundException.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/layer/LayerAlreadyExistsException.java  1
         

Complexity decreasing per file
==============================
+ modules/core/core-api/src/main/java/com/enonic/xp/project/ModifyProjectParams.java  -1
         

Clones added
============
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ModifyLayerParams.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/ContentLayer.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/layer/CreateLayerParams.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=5108750&bid=16799807)

**Pull request merged**